### PR TITLE
Validate sub-playbooks , scripts and integrations match the playbook version

### PR DESCRIPTION
**Pull request by abaumgarten**
## Status
- [ ] In Progress
- [x] Ready
- [ ] In Hold - (Reason for hold)

## Related Issues
fixes: https://github.com/demisto/etc/issues/23553

## Description
Validated all of the above: 
1. If a playbook uses an integration's command, validates that:  `integration_version <= playbook version.`
2. If a playbook uses a script, validates that:  `script_version <= playbook version.`
3. If a playbook uses a sub-playbook, validates that:  `sub_playbook_version <= playbook version.`

In order to validate the integration's command (point number 1), I have added a function that modifies the id_set file (after it was created). This function modifies `command_to_integration` data. For each command, the value was modified to be a list of all the implementing integrations instead of an empty string.  

Additional error handlers were added.
    
## Must have
- [x] Tests
- [x] Documentation

### CODE REVIEW

**Comment by abaumgarten** [comment]
I am working on adding UT, the PR is not ready for a review yet.

**Comment by abaumgarten** [comment]
@bakatzir @hod-alpert The PR is not ready for review yet, I have to add UTs

**Comment by hod-alpert** [comment]
@abaumgarten Can you also take a look at the long duration validate-files job takes?

It used to be ~13 minutes, now it's ~23 minutes and on your PR it's 43 minutes.

File: .circleci/config.yml [hunk 1]
```diff
@@ -162,7 +162,7 @@ jobs:
 
               export CIRCLE_ARTIFACTS="/home/circleci/project/artifacts"
 
-              gsutil cp gs://marketplace-dist/content/id_set.json ./Tests/id_set.json
+              CIRCLE_BRANCH="master" demisto-sdk create-id-set -o ./Tests/id_set.json
```
Comment (abaumgarten) [inline .circleci/config.yml:165]: It Will be deleted before merge

File: demisto_sdk/commands/common/hook_validations/id.py [hunk 2]
```diff
@@ -209,6 +212,152 @@ def _is_mapper_incident_types_found(self, mapper_data):
...
+                                                                                                  command,
+                                                                                                  playbook_version)
+                if self.handle_error(error_message, error_code, file_path):
+                    return False
+
+        return True
+
+    def get_integration_version(self, integration_to_search):
+        general_version = ""  # i.e integration has no specific version
+        for integration_dict in self.integration_set:
```
Comment (guykeller) [inline demisto_sdk/commands/common/hook_validations/id.py:354]: How big can the integrations set be? Can we make this code more efficient?

File: demisto_sdk/commands/common/update_id_set.py [hunk 3]
```diff
@@ -1396,12 +1395,6 @@ def re_create_id_set(id_set_path: Optional[str] = DEFAULT_ID_SET_PATH, pack_to_c
     new_ids_dict['Widgets'] = sort(widgets_list)
     new_ids_dict['Mappers'] = sort(mappers_list)
 
-    if id_set_path:
-        if not exists(id_set_path):
-            intermediate_dirs = os.path.dirname(os.path.abspath(id_set_path))
-            os.makedirs(intermediate_dirs, exist_ok=True)
-        with open(id_set_path, 'w+') as id_set_file:
-            json.dump(new_ids_dict, id_set_file, indent=4)
```
Comment (guykeller) [inline demisto_sdk/commands/common/update_id_set.py:1404]: Why is this removed?

File: demisto_sdk/commands/create_id_set/create_id_set.py [hunk 4]
```diff
@@ -14,6 +21,54 @@ def __init__(self, output: str = '', input: str = '', print_logs: bool = True):
...
         self.print_logs = print_logs
+        self.id_set = OrderedDict()  # type: ignore
 
     def create_id_set(self):
-        return re_create_id_set(id_set_path=self.output, pack_to_create=self.input, print_logs=self.print_logs)
+        self.id_set = re_create_id_set(id_set_path=self.output, pack_to_create=self.input, print_logs=self.print_logs)
+        self.modify_id_set_command_to_integration_of_playbook()
+        return self.id_set
+
+    def modify_id_set_command_to_integration_of_playbook(self):
```
Comment (guykeller) [inline demisto_sdk/commands/create_id_set/create_id_set.py:31]: WDYT? could be a different name too
```suggestion
    def add_command_to_implementing_integrations_mapping(self):
```

File: demisto_sdk/commands/create_id_set/create_id_set.py [hunk 5]
```diff
@@ -14,6 +21,54 @@ def __init__(self, output: str = '', input: str = '', print_logs: bool = True):
...
+            playbook_name = list(playbook_dict.keys())[0]
+            playbook_data = playbook_dict[playbook_name]
+            commands_to_integration = playbook_data.get("command_to_integration", {})
+            for command in commands_to_integration:
+                is_command_implemented_in_integration = command in command_name_to_implemented_integration_map
+                if is_command_implemented_in_integration:
+                    implemented_integration = command_name_to_implemented_integration_map[command]
+                    commands_to_integration[command] = implemented_integration
+
+        self.save_id_set()
```
Comment (guykeller) [inline demisto_sdk/commands/create_id_set/create_id_set.py:50]: I think this should happen in the `create_id_set`, don't you?
Contributes to the single responsibility principle

File: demisto_sdk/commands/create_id_set/create_id_set.py [hunk 6]
```diff
@@ -14,6 +21,54 @@ def __init__(self, output: str = '', input: str = '', print_logs: bool = True):
...
-        return re_create_id_set(id_set_path=self.output, pack_to_create=self.input, print_logs=self.print_logs)
+        self.id_set = re_create_id_set(id_set_path=self.output, pack_to_create=self.input, print_logs=self.print_logs)
+        self.modify_id_set_command_to_integration_of_playbook()
+        return self.id_set
+
+    def modify_id_set_command_to_integration_of_playbook(self):
+        """
+        Modifies playbook set in id_set dictionary once it was created, and finally saves it to a file.
+        Each playbook that has "command_to_integration" field will be modified :
+        - command name value will be a list of all integrations that implements this command (instead of use "" ).
```
Comment (guykeller) [inline demisto_sdk/commands/create_id_set/create_id_set.py:35]: Seems to me like we could be doing this more efficiently.
Could you explain why this is the most efficient way to achieve this in your opinion?

File: demisto_sdk/commands/create_id_set/create_id_set.py [hunk 7]
```diff
@@ -14,6 +21,54 @@ def __init__(self, output: str = '', input: str = '', print_logs: bool = True):
...
+            for command in commands_to_integration:
+                is_command_implemented_in_integration = command in command_name_to_implemented_integration_map
+                if is_command_implemented_in_integration:
+                    implemented_integration = command_name_to_implemented_integration_map[command]
+                    commands_to_integration[command] = implemented_integration
+
+        self.save_id_set()
+
+    def create_command_to_implemented_integration_map(self):
+        command_name_to_implemented_integration_map = {}  # type: ignore
```
Comment (guykeller) [inline demisto_sdk/commands/create_id_set/create_id_set.py:53]: Why was `# type: ignore` needed?

File: demisto_sdk/commands/create_id_set/create_id_set.py [hunk 8]
```diff
@@ -14,6 +21,54 @@ def __init__(self, output: str = '', input: str = '', print_logs: bool = True):
...
+        integrations_list = self.id_set['integrations']
+        for integration_dict in integrations_list:
```
Comment (guykeller) [inline demisto_sdk/commands/create_id_set/create_id_set.py:55]:
```suggestion
        for integration_data in integrations_list:
```

**Review by guykeller — CHANGES_REQUESTED** [review]

File: demisto_sdk/commands/create_id_set/create_id_set.py (hunk 6 above)
Comment (abaumgarten) [reply demisto_sdk/commands/create_id_set/create_id_set.py:35]: Unfortunately, I can't think of a better and more efficient idea for doing it. My goal is that each command used in the playbook will be mapped to all integrations that implemented it. To do that, I have to go over all integrations set and create such a mapping. I cannot see other solution that does not cost in go over all integrations set to create this mapping and at the same time satisfies my goal. As I wrote in previous comments, we can discuss changing this modify of id_set. WDYT?

**Review by guykeller — APPROVED** [review]
Please fix the conflict and build before merging

**Review by bakatzir — CHANGES_REQUESTED** [review]
As we use the id_set in the `collect tests` step in the content build, please trigger content nightly(`./Utils/trigger_content_nightly_build`) with this branch prior the merge. we don't care about the server TPB, but regarding the infra step that are using the id_set.

also, please run a profiler(py-spy) as the time of the `validate-files` step increased. let's make sure it is necessary.

File: demisto_sdk/commands/create_id_set/create_id_set.py [hunk 9]
```diff
@@ -1,7 +1,14 @@
+import json
+import os
+from collections import OrderedDict
+from genericpath import exists
+
 from demisto_sdk.commands.common.update_id_set import re_create_id_set
 
 
 class IDSetCreator:
+    DEFAULT_ID_SET_PATH = "./Tests/id_set.json"
```
[the remark this answers is not in the document]
Comment (abaumgarten) [reply demisto_sdk/commands/create_id_set/create_id_set.py]: Done. Moved it to constants.py and changed it in all relevant places as well.

**Comment by abaumgarten** [comment]
> As we use the id_set in the `collect tests` step in the content build, please trigger content nightly(`./Utils/trigger_content_nightly_build`) with this branch prior the merge. we don't care about the server TPB, but regarding the infra step that are using the id_set.
> 
> also, please run a profiler(py-spy) as the time of the `validate-files` step increased. let's make sure it is necessary.

I ran those 2 nightly builds:
1. on content: https://app.circleci.com/pipelines/github/demisto/content/65139/workflows/09d90841-0957-4570-b057-d6370c1500d3
2. SDK nightly: https://app.circleci.com/pipelines/github/demisto/content/65096/workflows/45d48b95-770d-48d5-ba6f-23e46cfcf963

**Comment by bakatzir** [comment]
@abaumgarten great. please fix in the content repo tha faulty PBs so when we will merge this the build will not break.
```
=========== Found errors in the following files ===========

Packs/EmployeeOffboarding/Playbooks/Employee_Offboarding_-_Gather_User_Information.yml - [PB111]
Packs/GoogleChronicleBackstory/Playbooks/playbook-ChronicleAsset_Investigation_-_Chronicle.yml - [PB110]
Packs/GoogleChronicleBackstory/Playbooks/playbook-Hostname_And_IP_Address_Investigation_And_Remediation_-_Chronicle.yml - [PB111]
Packs/PrismaAccess/Playbooks/playbook-Prisma_Access_-__Logout_User.yml - [PB111]
Packs/TIM_SIEM/Playbooks/playbook-TIM_-_Add_All_Indicator_Types_To_SIEM.yml - [PB110]
```

**Comment by abaumgarten** [comment]
> @abaumgarten great. please fix in the content repo tha faulty PBs so when we will merge this the build will not break.
> 
> ```
> =========== Found errors in the following files ===========
> 
> Packs/EmployeeOffboarding/Playbooks/Employee_Offboarding_-_Gather_User_Information.yml - [PB111]
> Packs/GoogleChronicleBackstory/Playbooks/playbook-ChronicleAsset_Investigation_-_Chronicle.yml - [PB110]
> Packs/GoogleChronicleBackstory/Playbooks/playbook-Hostname_And_IP_Address_Investigation_And_Remediation_-_Chronicle.yml - [PB111]
> Packs/PrismaAccess/Playbooks/playbook-Prisma_Access_-__Logout_User.yml - [PB111]
> Packs/TIM_SIEM/Playbooks/playbook-TIM_-_Add_All_Indicator_Types_To_SIEM.yml - [PB110]
> ```
Already fixed here (waiting to @yaakovi to approve and force merge):
https://github.com/demisto/content/pull/11114

**Review by bakatzir — APPROVED** [review]

**Comment by bakatzir** [comment]
@yaakovi We had multiple build issues this week, I suggest be heavy headed on force merging. demisto/content#11114

**Comment by abaumgarten** [comment]
@hod-alpert I reverted the build changes, so we will have to force the merge?

**Comment by ChanochShayner** [comment]
This PR probably break the SDK:
Because validate runs against ID set from the bucket (and this PR modified the ID set)
my opinion is to use create ID set instead of downloading it from the bucket (at least for now)
@roysagi @bakatzir @hod-alpert what do you think?
thanks!

**Comment by abaumgarten** [comment]
update: we will wait with the merge to the next SDK release to avoid reverting changes in the build. 
@hod-alpert @bakatzir @roysagi @yuvalbenshalom  FYI
